src/nlp/features.py

```diff
@@ -95,6 +95,115 @@ def __call__(self):
         return self.pa_type
 
 
+@dataclass
+class MultiArray:
+
+    shape: Union[Tuple[int], List[int]]
+    dtype: str  # innermost element dtype
+    id: Optional[str] = None
+
+    def __new__(cls, shape, dtype):
+        array_class = MultiArray._extension_array_factory(shape)
+        extension_class = MultiArray._extension_type_factory(shape, dtype, array_class)
+        return extension_class()
+
+    @staticmethod
+    def _generate_dtype(shape, ndims, dtype, current_dim=0):
+        if current_dim == ndims:
+            return dtype
+        elif current_dim == 0:
+            dtype = pa.list_(string_to_arrow(dtype), shape[-1])
+        elif current_dim < ndims:
+            dtype = pa.list_(dtype)
+        current_dim += 1
+        return MultiArray._generate_dtype(shape, ndims, dtype, current_dim)
+
+    @staticmethod
+    def _generate_flatten(ndims, storage, current_dim=0):
+        if current_dim == ndims:
+            return storage.to_numpy()
+        elif current_dim < ndims:
+            storage = storage.flatten()
+        current_dim += 1
+        return MultiArray._generate_flatten(ndims, storage, current_dim)
+
+    @staticmethod
+    def _extension_type_factory(shape, dtype, array_type):
+
+        extension_name = f'TypeArray{len(shape)}D'
+        if extension_name in globals():
+            return eval(extension_name)
+
+        def _custom_init(self):
+            pa.PyExtensionType.__init__(self, self.storage_type)
+
+        def _custom_reduce(self):
+            return self.__class__, ()
+
+        def _custom_arrow_ext(self):
+            return array_type
+
+        def __call__(self):
+            return self
+
+        def encode_example(self, value):
+            assert isinstance(value, numpy.ndarray)
+            extension_type = eval(f'TypeArray{len(value.shape)}D')
+            value_dtype = MultiArray._generate_dtype(shape, len(shape), str(value.dtype))
+            value = value[numpy.newaxis, ...].tolist()
+            encoded = pa.ExtensionArray.from_storage(
+                extension_type(),
+                pa.array(value, value_dtype))
+            return encoded
+
+        factory = type(
+            extension_name,
+            (pa.PyExtensionType,),
+            {
+                '__init__': _custom_init,
+                '__arrow_ext_class__': _custom_arrow_ext,
+                '__reduce__': _custom_reduce,
+                '__call__': __call__,
+                'storage_type': MultiArray._generate_dtype(shape, len(shape), dtype),
+                'encode_example': encode_example,
+                'extension_name': extension_name
+            }
+        )
+        globals()[extension_name] = factory
```

Comment:
What if we instantiate two `TypeArray2D` classes in a row, will the second one overwrite the first one, and therefore delete the information about the shape ?

Reply:
We definitely could just re-instantiate the class, since any array of a particular dimension will become invariant of the shape. I added this little bit just so as not to redefine the same class if that particular dim-array class had already been instantiated. I'll go ahead and remove it since it shouldn't matter if we define specific dimension array classes later